sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiLoads.java

```diff
@@ -52,16 +52,18 @@
 /** This {@link PTransform} manages loads into BigQuery using the Storage API. */
 public class StorageApiLoads<DestinationT, ElementT>
     extends PTransform<PCollection<KV<DestinationT, ElementT>>, WriteResult> {
-  final TupleTag<KV<DestinationT, StorageApiWritePayload>> successfulConvertedRowsTag =
-      new TupleTag<>("successfulRows");
+  final TupleTag<KV<DestinationT, KV<ElementT, StorageApiWritePayload>>>
```

Comment:
Here it changed PTransform output element type. Do we need some change in BigQueryTranslation to preserve upgrade compatibility? cc: @chamikaramj 

or is there plan to setup precommit test for bigquery pipeline upgrade? so tests can auto detect this (like kafka upgrade project)

Reply:
I think changing output element type and the coder here could break streaming update compatibility in general.

Have you considered using the `updateCompatibilityVersion` option ?

https://github.com/apache/beam/blob/c531f898d6b3454761f78a2420df5424178fb002/sdks/java/core/src/main/java/org/apache/beam/sdk/options/StreamingOptions.java#L45

Reply:
That would require us to maintain two implementations of Streaming Inserts, one with this change, and one without, right? I think that would be prohibitive in general for beam IOs

Reply:
This should be called out in CHANGES.md if we have to do these breaking changes. But I recommend `updateCompatibilityVersion` here.

Reply:
Previously, we also pushed such compatibility breaking changes behind switches in the API (for example, `withClustering`). But I think `updateCompatibilityVersion` nicer.

Reply:
There are different ways to flag it to be sure, but we would still need to maintain two versions, one that supported the KV version, and one that supports the old version, right?

Reply:
Correct, unfortunately we need to keep the old version around if we want to maintain streaming update compatibility I believe.

Reply:
Is that something we actually want to do? That would imply that we will eventually be maintaining a large number of versions for compat reasons. That doesn't sound like its feasible to me.

Reply:
Given the heavy usage of this IO, I think we should maintain streaming update compatibility.

Reply:
Ok. I think we need a better strategy than simply manually maintaining every version then

Reply:
Also, is there a test that can be written to verify this behavior?

Reply:
So, should I just copy paste the classes?

Reply:
I think we should have a broader version on updateCompatibilyVersion, because this looks like we will, over time, start accruing legacy code based on specific beam versions. That strikes me as eventually unmaintainable.

Reply:
Because we will need to do this whenever we change the shape of a PCollection

Reply:
I think one way to do that might be to use schema'd `PCollection`s everywhere and evolve the schema when we want to do changes to the `PCollection` structure. But for the time being we have to fork the code to preserve compatibility. BTW not all structure changes will break compatibility so some changes can be done without forking (Dataflow link above has some details) but generally coder changes can break update compatibility (and probably Dataflow compatibility check will fail).

Reply:
Ok. I'll make the forking change for this when I have a chance

Reply:
If we encounter an unavoidable update-incompatible change in the future, can we revert these extra files? I share @johnjcasey's aversion towards duplicating substantial and complex code.

Reply:
Yeah, this looks pretty bad.

I wonder if instead of duplicating code here we can introduce an option to BQ (say 'withLegacyOutputFormat' or something more descriptive) that results in elements being pushed to a new tag with the old format (say 'successfulWrittenRowsTagWithoutElement') with the old coder.

That way, anyone who needs update compatibility can use that option and we do not need to fork code.

Reply:
cc: @scwhittle @kennknowles @robertwb in case they have ideas regarding maintaining update compatibility without a lot of code duplication here.

Reply:
I haven't dug into the code too much. Can you write something generic that is instantiated in both code paths? Having two update compatibility paths doesn't necessarily mean duplicating all the code.

Reply:
My basic problem is that I need to carry a new value across the ptransform, which basically requires a new coder. In this case, the type of the PTransform is changed as well, so I'm not entirely certain how I could avoid much of this duplication

Reply:
Aside from update compatibility issues, doesn't this increase the data shuffled as we are now shuffling the writepayload and the original elements?  If so it seems that we might want the previous behavior not just for older SDKs but also in cases where an error function requiring the original element is not configured.

Do we need to change the output type for successful writes? It seems like the original element is just being added for error handling path.

@reuvenlax 

If I understand correctly the graph is not changing here, just the encoding of the elements and we're going from StorageApiWritePayload to KV<E, StorageApiWritePayload>.  Would it be possible to have a special coder for KV<E, StorageApiWritePayload> such that it handles decoding previously coded StorageApiWritePayload as KV<null, StorageApiWritePayload>?
It seems like that could be done via a Schema in some way since StorageApiWritePayload uses autovalue schema coder.  I think the dataflow backend would note that the schema is compatible in that case and allow the update to proceed.

Or a simpler route, could the element just be added as a nullable field to StorageApiWritePayload instead of changing to KV<E, StorageApiWritePayload>?

To share code, could we just switch to the new type throughout and have the element be null if not needed or missing due to previously encoded? Since the new type is a superset of the old, it seems like the compatibility with previous sdks could be kept to the boundaries of the impl (if the above doesn't let you share completely).

Reply:
The idea of a coder that handles both decodings is very interesting to me. We could use the compat flag to indicate using the old coder, instead of the new coder, which would preserve backwards compatibility much better. I'll try that, as I think thats a good way to do this

Reply:
this wouldn't even require the update compat flag i think actually

Reply:
As a side comment, this is another motivation to use schema coders more ubiquitously--adding another field is update compatible. 

On another note, anything that involves shuffling more data in the main data path should be looked at carefully from a perf standpoint. We've gone to a lot of effort (e.g. with dynamic destinations) to ensure shuffling metadata doesn't become a perf impediment.